<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Changelog • pcadapt</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-news">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pcadapt</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">4.3.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/pcadapt.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/bcm-uga/pcadapt/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small></small></h1>
      <small>Source: <a href="https://github.com/bcm-uga/pcadapt/blob/HEAD/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="page-header" data-toc-text="4.0" id="pcadapt-40">pcadapt 4.0<a class="anchor" aria-label="anchor" href="#pcadapt-40"></a></h2>
<ul><li><p><code><a href="../reference/read.pcadapt.html">read.pcadapt()</a></code> generates <code>bed</code> files instead of <code>pcadapt</code> files.</p></li>
<li><p>Computation of PCA is now based on R package <code>RSpectra</code>.</p></li>
<li><p>Missing values are handled by specifying matrix-vector operations in <code>RSpectra</code> that accounts for missing values.</p></li>
<li><p>Includes LD thinning to compute PCs.</p></li>
<li><p>No more dependency to R package RcppArmadillo.</p></li>
<li><p>For Pooled-seq data, use Mahalanobis distances based on PCA loadings, no more simulations of individual genotypes.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="3.1" id="pcadapt-31">pcadapt 3.1<a class="anchor" aria-label="anchor" href="#pcadapt-31"></a></h2>
<ul><li><p>Switch from C/Lapack to Rcpp/RcppArmadillo.</p></li>
<li><p><code><a href="../reference/pcadapt.html">pcadapt()</a></code> can take genotype matrices as input.</p></li>
<li><p>Modified code for binomial sampling.</p></li>
<li><p><code><a href="../reference/pcadapt.html">pcadapt()</a></code> argument <code>clean.files</code> is now deprecated.</p></li>
<li><p><code><a href="../reference/pcadapt.html">pcadapt()</a></code> argument <code>output.filename</code> is now deprecated.</p></li>
<li><p><code><a href="../reference/read.pcadapt.html">read.pcadapt()</a></code> argument <code>local.env</code> is now deprecated.</p></li>
<li><p>Latest update of R package <code>vcfR</code> taken into account.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="3.0.4" id="pcadapt-304">pcadapt 3.0.4<small>2017-01-02</small><a class="anchor" aria-label="anchor" href="#pcadapt-304"></a></h2>
<ul><li>Method based on sampling genotypes added to handle pooled-sequencing.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="3.0.3" id="pcadapt-303">pcadapt 3.0.3<small>2016-11-03</small><a class="anchor" aria-label="anchor" href="#pcadapt-303"></a></h2>
<ul><li><p>Option <code>type = "vcfR"</code> has been added to <code><a href="../reference/read.pcadapt.html">read.pcadapt()</a></code> to overcome some conversion issues occurring with VCF files.</p></li>
<li><p>Argument <code>transpose</code> is now deprecated. Read section A for more details.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="3.0.2" id="pcadapt-302">pcadapt 3.0.2<small>2016-06-21</small><a class="anchor" aria-label="anchor" href="#pcadapt-302"></a></h2>
<ul><li>The function <code><a href="../reference/get.pc.html">get.pc()</a></code> has been added. For each SNP, it returns the most correlated principal component.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="3.0.1" id="pcadapt-301">pcadapt 3.0.1<small>2016-06-14</small><a class="anchor" aria-label="anchor" href="#pcadapt-301"></a></h2>
<ul><li><p>Function <code>read4pcadapt()</code> is now deprecated, it is now called <code><a href="../reference/read.pcadapt.html">read.pcadapt()</a></code>.</p></li>
<li><p>Using the <code>pop</code> option when plotting scores now provides the color legend.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="3.0" id="pcadapt-30">pcadapt 3.0<small>2016-04-15</small><a class="anchor" aria-label="anchor" href="#pcadapt-30"></a></h2>
<ul><li><p>All analyses are now included in the R package. Users should not use the C software PCAdapt fast anymore.</p></li>
<li><p>Big datasets can be handled directly within the R session.</p></li>
<li><p><code>read4pcadapt()</code> now converts files to the <code>pcadapt</code> format.</p></li>
<li><p>The first argument of <code><a href="../reference/pcadapt.html">pcadapt()</a></code> can be either a small genotype matrix or the output of <code>read4pcadapt()</code>.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="2.2" id="pcadapt-22">pcadapt 2.2<small>2016-02-13</small><a class="anchor" aria-label="anchor" href="#pcadapt-22"></a></h2>
<ul><li><p>The Mahalanobis distance is now estimated from the z-scores rather than the loadings.</p></li>
<li><p>Make sure you have downloaded the latest version of the C software PCAdapt (last updated on February 11, 2016).</p></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="2.1" id="pcadapt-21">pcadapt 2.1<small>2015-12-18</small><a class="anchor" aria-label="anchor" href="#pcadapt-21"></a></h2>
<ul><li><p>The scaling of the SNP before computing PCA has been changed. Instead of using standard deviation, we now use the square root of <span class="math inline"><em>p</em>(1−<em>p</em>)</span> (haploid data) or of <span class="math inline">2<em>p</em>(1−<em>p</em>)</span> (diploid data) where <span class="math inline"><em>p</em></span> is the minimum allele frequency.</p></li>
<li><p>Bug fix: the genomic inflation factor has been corrected when <code>K=1</code>.</p></li>
<li><p>Bug fix: a problem due to high proportion of missing data slowing the program has been fixed.</p></li>
<li><p>Argument <code>"minmaf"</code> has been replaced with <code>"min.maf"</code>.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="2.0" id="pcadapt-20">pcadapt 2.0<small>2015-11-04</small><a class="anchor" aria-label="anchor" href="#pcadapt-20"></a></h2>
<ul><li><p>The default test statistic is not the communality statistic anymore but the Mahalanobis distance.</p></li>
<li><p>Test statistic for Pool-seq data.</p></li>
</ul></div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>

</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Keurcien Luu, Michael Blum, Florian Privé.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

