<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Principal Component Analysis for outlier detection — pcadapt • pcadapt</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Principal Component Analysis for outlier detection — pcadapt"><meta property="og:description" content="pcadapt performs principal component analysis and computes p-values to
test for outliers. The test for outliers is based on the correlations between
genetic variation and the first K principal components. pcadapt
also handles Pool-seq data for which the statistical analysis is performed on
the genetic markers frequencies. Returns an object of class pcadapt."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pcadapt</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">4.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/pcadapt.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/bcm-uga/pcadapt/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Principal Component Analysis for outlier detection</h1>
    <small class="dont-index">Source: <a href="https://github.com/bcm-uga/pcadapt/blob/HEAD/R/pcadapt.R" class="external-link"><code>R/pcadapt.R</code></a></small>
    <div class="hidden name"><code>pcadapt.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>pcadapt</code> performs principal component analysis and computes p-values to
test for outliers. The test for outliers is based on the correlations between
genetic variation and the first <code>K</code> principal components. <code>pcadapt</code>
also handles Pool-seq data for which the statistical analysis is performed on
the genetic markers frequencies. Returns an object of class <code>pcadapt</code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">pcadapt</span><span class="op">(</span></span>
<span>  <span class="va">input</span>,</span>
<span>  K <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"mahalanobis"</span>,</span>
<span>  min.maf <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  ploidy <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  LD.clumping <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pca.only <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  tol <span class="op">=</span> <span class="fl">1e-04</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for pcadapt_matrix</span></span>
<span><span class="fu">pcadapt</span><span class="op">(</span></span>
<span>  <span class="va">input</span>,</span>
<span>  K <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mahalanobis"</span>, <span class="st">"componentwise"</span><span class="op">)</span>,</span>
<span>  min.maf <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  ploidy <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  LD.clumping <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pca.only <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  tol <span class="op">=</span> <span class="fl">1e-04</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for pcadapt_bed</span></span>
<span><span class="fu">pcadapt</span><span class="op">(</span></span>
<span>  <span class="va">input</span>,</span>
<span>  K <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mahalanobis"</span>, <span class="st">"componentwise"</span><span class="op">)</span>,</span>
<span>  min.maf <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  ploidy <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  LD.clumping <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pca.only <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  tol <span class="op">=</span> <span class="fl">1e-04</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for pcadapt_pool</span></span>
<span><span class="fu">pcadapt</span><span class="op">(</span></span>
<span>  <span class="va">input</span>,</span>
<span>  K <span class="op">=</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">input</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"mahalanobis"</span>,</span>
<span>  min.maf <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  ploidy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  LD.clumping <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pca.only <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">tol</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>input</dt>
<dd><p>The output of function <code>read.pcadapt</code>.</p></dd>


<dt>K</dt>
<dd><p>an integer specifying the number of principal components to retain.</p></dd>


<dt>method</dt>
<dd><p>a character string specifying the method to be used to compute
the p-values. Two statistics are currently available, <code>"mahalanobis"</code>,
and <code>"componentwise"</code>.</p></dd>


<dt>min.maf</dt>
<dd><p>Threshold of minor allele frequencies above which p-values are 
computed. Default is <code>0.05</code>.</p></dd>


<dt>ploidy</dt>
<dd><p>Number of trials, parameter of the binomial distribution. 
Default is 2, which corresponds to diploidy, such as for the human genome.</p></dd>


<dt>LD.clumping</dt>
<dd><p>Default is <code>NULL</code> and doesn't use any SNP thinning.
If you want to use SNP thinning, provide a named list with parameters 
<code>$size</code> and <code>$thr</code> which corresponds respectively to the window 
radius and the squared correlation threshold. A good default value would 
be <code>list(size = 500, thr = 0.1)</code>.</p></dd>


<dt>pca.only</dt>
<dd><p>a logical value indicating whether PCA results should be 
returned (before computing any statistic).</p></dd>


<dt>tol</dt>
<dd><p>Convergence criterion of <code><a href="https://rdrr.io/pkg/RSpectra/man/svds.html" class="external-link">RSpectra::svds()</a></code>. 
Default is <code>1e-4</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>The returned value is an object of class <code>pcadapt</code>.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>First, a principal component analysis is performed on the scaled and 
centered genotype data. Depending on the specified <code>method</code>, different 
test  statistics can be used.</p>
<p><code>mahalanobis</code> (default): the robust Mahalanobis distance is computed for 
each genetic marker using a robust estimate of both mean and covariance 
matrix between the <code>K</code> vectors of z-scores.</p>
<p><code>communality</code>: the communality statistic measures the proportion of 
variance explained by the first <code>K</code> PCs. Deprecated in version 4.0.0.</p>
<p><code>componentwise</code>: returns a matrix of z-scores.</p>
<p>To compute p-values, test statistics (<code>stat</code>) are divided by a genomic 
inflation factor (<code>gif</code>) when <code>method="mahalanobis"</code>. When using 
<code>method="mahalanobis"</code>, the scaled statistics 
(<code>chi2_stat</code>) should follow a chi-squared distribution with <code>K</code> 
degrees of freedom. When using <code>method="componentwise"</code>, the z-scores 
should follow a chi-squared distribution with <code>1</code> degree of freedom. For 
Pool-seq data, <code>pcadapt</code> provides p-values based on the Mahalanobis 
distance for each SNP.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Keurcien Luu, Michael Blum, Florian Privé.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

