# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

get_af <- function(obj) {
    .Call(`_pcadapt_get_af`, obj)
}

bedXPtr <- function(path, n, p) {
    .Call(`_pcadapt_bedXPtr`, path, n, p)
}

bed2mat <- function(xptr) {
    .Call(`_pcadapt_bed2mat`, xptr)
}

clumping <- function(obj, ind_col, ord, remain, size, thr) {
    .Call(`_pcadapt_clumping`, obj, ind_col, ord, remain, size, thr)
}

multLinReg <- function(obj, ind_col, af, ploidy, u) {
    .Call(`_pcadapt_multLinReg`, obj, ind_col, af, ploidy, u)
}

nb_nona <- function(obj, ind_col) {
    .Call(`_pcadapt_nb_nona`, obj, ind_col)
}

#' Summary
#'
#' \code{print_convert} prints out a summary of the file conversion.
#'
#' @param input a genotype matrix or a character string specifying the name of the file to be converted.
#' @param output a character string specifying the name of the output file.
#' @param M an integer specifying the number of genetic markers present in the data.
#' @param N an integer specifying the number of individuals present in the data.
#' @param pool an integer specifying the type of data. `0` for genotype data, `1` for pooled data.
#'
#' @keywords internal
#'
print_convert <- function(input, output, M, N, pool) {
    invisible(.Call(`_pcadapt_print_convert`, input, output, M, N, pool))
}

#' Convert ped files
#'
#' \code{ped2pcadapt} converts \code{ped} files to the format \code{pcadapt}.
#'
#' @param input a character string specifying the name of the file to be converted.
#' @param output a character string specifying the name of the output file.
#' 
#' @keywords internal
#'
ped2pcadapt <- function(input, output) {
    .Call(`_pcadapt_ped2pcadapt`, input, output)
}

pMatVec4 <- function(obj, ind_col, af, ploidy, x) {
    .Call(`_pcadapt_pMatVec4`, obj, ind_col, af, ploidy, x)
}

cpMatVec4 <- function(obj, ind_col, af, ploidy, x) {
    .Call(`_pcadapt_cpMatVec4`, obj, ind_col, af, ploidy, x)
}

total_var_scaled <- function(obj, ind_col, af, ploidy) {
    .Call(`_pcadapt_total_var_scaled`, obj, ind_col, af, ploidy)
}

#' Convert vcfR genotype matrices
#'
#' \code{vcf_convert} converts outputs of \code{extract.gt} to the format 
#' \code{pcadapt}.
#'
#' @param string_geno a genotype matrix extracted from a VCF file with `vcfR`. 
#' @param output a character string indicating the name of the output file.
#' @param allele_sep a vector of characters indicating what delimiters are used 
#' to separate alleles.
#'
#' @keywords internal
#'
vcf_convert <- function(string_geno, output, allele_sep) {
    .Call(`_pcadapt_vcf_convert`, string_geno, output, allele_sep)
}

writebed <- function(filename, e, tab, is_pcadapt) {
    invisible(.Call(`_pcadapt_writebed`, filename, e, tab, is_pcadapt))
}

